woocommerce/qit-cli
# Custom Tests

## Custom Tests

This PR allows Partner developers of the Woo marketplace to submit their own E2E tests to QIT, and to run the E2E tests that we make available to them.

### Testing instructions

- Spin up the Manager on the custom tests branch
- Go to `_tests/custom_tests` and run `composer install`
- Copy `_tests/custom_tests/.env.sample` to `_tests/custom_tests/.env`
- Fill out the `.env` with the necessary values
- Run `./vendor/bin/paratest`, it should take around 5 minutes and pass
- This provides the coverage outlined [here](https://github.com/woocommerce/qit-cli/issues/162)

### Manual testing instructions

Explore the new commands added, which are:

- qit scaffold:e2e
- qit run:e2e
- qit tag:upload
- qit tag:delete
- qit tag:list

### Scaffold a test, upload it, list it and run it:

- `qit scaffold:e2e /tmp/foo`
- `qit tag:upload automatewoo /tmp/foo`
- `qit tag:list`
- `qit run:e2e automatewoo --plugin woocommerce:activate`

### Publish another test tag

- Now scaffold another test and publish it for `automatewoo-birthdays`
- Run it with `qit run:e2e automatewoo-birthdays --plugin woocommerce:activate --plugin automatewoo:test`
- You can also do this using a `qit-env.yml` or `qit-env.json` files

There are many options, try out with `--php_version 8.3`, different WP and Woo versions, object cache, etc.

Also try out  `--codegen` and `--ui`.

The self-tests is pretty self-explanatory too, if you're interested. Check [this](https://github.com/woocommerce/qit-cli/blob/14769aec80f6fa3305e11c8ba3ee80f4a849d3a1/_tests/custom_tests/tests/RunE2ETest.php) out for instance:

```php
	public function test_runs_scaffolded_e2e() {
		$output = qit( [
				'run:e2e',
				'automatewoo',
				$this->scaffold_test(),
				'--plugin',
				'woocommerce:activate',
			]
		);

		$output = $this->normalize_scaffolded_test_run_output( $output );

		$this->assertMatchesNormalizedSnapshot( $output );
	}
```

### Documentation:

The documentation contains example commands that we can use to try more scenarios: https://qit.woo.com/docs/custom-tests/introduction
